# Проект спринта 7. Секреты темнолесья. Предобработка данных в Python.

- Автор: Хазанов Михаил
- Дата: 30.01.2025

### Цели и задачи проекта

<font color='#777778'>Перечислите цель проекта и задачи, которые вы решаете. Для этого можно использовать описание проекта, но полезнее будет сформулировать основную цель проекта самостоятельно.</font>

Цель - помочь коллегам из «Секретов Темнолесья» изучить развитие игровой индустрии.
Задачи - ознакомиться с данными, проверить их корректность и провести предобработку, получив необходимый срез данных.
Перед анализом необходимо дополнительно сделать следующее:
* Отобрать данные по времени выхода игры. Вам нужен период с 2000 по 2013 год включительно.
* Категоризовать игры по оценкам пользователей и экспертов. Выделите три категории:
-- высокая оценка — с оценкой от 8 до 10 и от 80 до 100, включая правые границы интервалов.
-- средняя оценка — с оценкой от 3 до 8 и от 30 до 80, не включая правые границы интервалов.
-- низкая оценка — с оценкой от 0 до 3 и от 0 до 30, не включая правые границы интервалов.
* Выделить топ-7 платформ по количеству игр, выпущенных за весь требуемый период.

### Содержимое проекта

<font color='#777778'>Перечислите основные шаги проекта или напишите оглавление. Используйте описание проекта, чтобы зафиксировать основные шаги.</font>

* Загрузка данных и знакомство с ними.
* Проверка ошибок в данных и их предобработка.
* Фильтрация данных.
* Категоризация данных.
* Итоговый вывод.

---

## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [1]:
# Пустые ячейки после каждого задания — примерное пространство для работы.
# Вы можете свободно добавлять или удалять ячейки по своему усмотрению в зависимости от логики и объёма работы.

# Импортируем библиотеку pandas
import pandas as pd

# Выгружаем hotel_dataset.csv в датафрейм
df = pd.read_csv(...)



In [2]:
df

Name Platform  Year of Release         Genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16952               LMA Manager 2007     X360           2006.0        Sports   
16953        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16954               Spirits & Spells      GBA           2003.0      Platform   
16955            Winning Post 8 2016      PSV           2016.0    Simulation   

       NA sales EU sales JP sales  Other sales  Critic Score User Score Rating  
0         41.36    28.96     3.77         8.45          76.0          8      E  
1         29.08     3.58     6.81         0.77           NaN        NaN    NaN  
2         15.68    12.76     3.79         3.29          82.0        8.3      E  
3         15.61    10.93     3.28         2.95          80.0          8      E  
4         11.27     8.89    10.22         1.00           NaN        NaN    NaN  
...         ...      ...      ...          ...           ...        ...    ...  
16951      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16952      0.00     0.01      0.0         0.00           NaN        NaN    NaN  
16953      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16954      0.01      0.0      0.0         0.00           NaN        NaN    NaN  
16955      0.00      0.0     0.01         0.00           NaN        NaN    NaN  

[16956 rows x 11 columns]

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [3]:

# Смотрим общую информацию, ненулевые строчки
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [4]:

# Смотрим общую информацию, ненулевые строчки
df.head(15)


Name Platform  Year of Release         Genre  \
0                    Wii Sports      Wii           2006.0        Sports   
1             Super Mario Bros.      NES           1985.0      Platform   
2                Mario Kart Wii      Wii           2008.0        Racing   
3             Wii Sports Resort      Wii           2009.0        Sports   
4      Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                        Tetris       GB           1989.0        Puzzle   
6         New Super Mario Bros.       DS           2006.0      Platform   
7                      Wii Play      Wii           2006.0          Misc   
8     New Super Mario Bros. Wii      Wii           2009.0      Platform   
9                     Duck Hunt      NES           1984.0       Shooter   
10                   Nintendogs       DS           2005.0    Simulation   
11                Mario Kart DS       DS           2005.0        Racing   
12  Pokemon Gold/Pokemon Silver       GB           1999.0  Role-Playing   
13                      Wii Fit      Wii           2007.0        Sports   
14           Kinect Adventures!     X360           2010.0          Misc   

    NA sales EU sales JP sales  Other sales  Critic Score User Score Rating  
0      41.36    28.96     3.77         8.45          76.0          8      E  
1      29.08     3.58     6.81         0.77           NaN        NaN    NaN  
2      15.68    12.76     3.79         3.29          82.0        8.3      E  
3      15.61    10.93     3.28         2.95          80.0          8      E  
4      11.27     8.89    10.22         1.00           NaN        NaN    NaN  
5      23.20     2.26     4.22         0.58           NaN        NaN    NaN  
6      11.28     9.14      6.5         2.88          89.0        8.5      E  
7      13.96     9.18     2.93         2.84          58.0        6.6      E  
8      14.44     6.94      4.7         2.24          87.0        8.4      E  
9      26.93     0.63     0.28         0.47           NaN        NaN    NaN  
10      9.05    10.95     1.93         2.74           NaN        NaN    NaN  
11      9.71     7.47     4.13         1.90          91.0        8.6      E  
12      9.00     6.18      7.2         0.71           NaN        NaN    NaN  
13      8.92     8.03      3.6         2.15          80.0        7.7      E  
14     15.00     4.89     0.24         1.69          61.0        6.3      E

In [5]:
# Подсчитываем количество пропусков в каждом столбце
missing_counts = df.isna().sum()

# Рассчитываем долю пропусков
missing_part = missing_counts / len(df)

# Создаем датафрейм с результатами
missing_data = pd.DataFrame({'Количество пропусков': missing_counts, 'Доля пропусков': missing_part})
missing_data


Количество пропусков  Доля пропусков
Name                                2        0.000118
Platform                            0        0.000000
Year of Release                   275        0.016218
Genre                               2        0.000118
NA sales                            0        0.000000
EU sales                            0        0.000000
JP sales                            0        0.000000
Other sales                         0        0.000000
Critic Score                     8714        0.513918
User Score                       6804        0.401274
Rating                           6871        0.405225

- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

<font color='#777778'>Используйте ячейки типа Markdown для промежуточных выводов и расширенных комментариев к действиям с данными. </font>

В наших данных 16956 строк и 11 столбцов. 4 столбца находятся в вещественном формате и 7 в формате object. Самое большое количество пропусков хранится в столбцах 'Critic Score' - 51.4%, 'User Score' - 40.1%, 'Rating' - 40.5%. В типе данных год надо привести к целочисленному значению int, оценку пользователя - к float


Названия столбцов написаны корректно (для меня). Все на английском языке, при необходимости можем перевести все на русский.

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [6]:

# Выводим названия всех столбцов
column_all = df.columns.tolist()
column_all

['Name',
 'Platform',
 'Year of Release',
 'Genre',
 'NA sales',
 'EU sales',
 'JP sales',
 'Other sales',
 'Critic Score',
 'User Score',
 'Rating']

In [7]:


# Приведение названий столбцов к стилю snake case
# Заменяем пробелы на нижние подчеркивания в названиях столбцов
df.columns = df.columns.str.replace(' ', '_')

# Приводим все названия столбцов к нижнему регистру
df.columns = df.columns.str.lower()

# Выводим обновленные названия столбцов
new_columns = df.columns.tolist()
new_columns


['name',
 'platform',
 'year_of_release',
 'genre',
 'na_sales',
 'eu_sales',
 'jp_sales',
 'other_sales',
 'critic_score',
 'user_score',
 'rating']

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

In [8]:
# Уникальные значения по столбцам с несоответствием
unique_eu_sales = df['eu_sales'].unique()
unique_jp_sales = df['jp_sales'].unique()
unique_user_score = df['user_score'].unique()
unique_critic_score = df['critic_score'].unique()
unique_years = df['year_of_release'].unique()
unique_rating = df['rating'].unique()

display(unique_eu_sales)
display(unique_jp_sales)
display(unique_user_score)
display(unique_critic_score)
display(unique_years)
display(unique_rating)

array(['28.96', '3.58', '12.76', '10.93', '8.89', '2.26', '9.14', '9.18',
       '6.94', '0.63', '10.95', '7.47', '6.18', '8.03', '4.89', '8.49',
       '9.09', '0.4', '3.75', '9.2', '4.46', '2.71', '3.44', '5.14',
       '5.49', '3.9', '5.35', '3.17', '5.09', '4.24', '5.04', '5.86',
       '3.68', '4.19', '5.73', '3.59', '4.51', '2.55', '4.02', '4.37',
       '6.31', '3.45', '2.81', '2.85', '3.49', '0.01', '3.35', '2.04',
       '3.07', '3.87', '3.0', '4.82', '3.64', '2.15', '3.69', '2.65',
       '2.56', '3.11', '3.14', '1.94', '1.95', '2.47', '2.28', '3.42',
       '3.63', '2.36', '1.71', '1.85', '2.79', '1.24', '6.12', '1.53',
       '3.47', '2.24', '5.01', '2.01', '1.72', '2.07', '6.42', '3.86',
       '0.45', '3.48', '1.89', '5.75', '2.17', '1.37', '2.35', '1.18',
       '2.11', '1.88', '2.83', '2.99', '2.89', '3.27', '2.22', '2.14',
       '1.45', '1.75', '1.04', '1.77', '3.02', '2.75', '2.16', '1.9',
       '2.59', '2.2', '4.3', '0.93', '2.53', '2.52', '1.79', '1.3', '2.6',
   

array(['3.77', '6.81', '3.79', '3.28', '10.22', '4.22', '6.5', '2.93',
       '4.7', '0.28', '1.93', '4.13', '7.2', '3.6', '0.24', '2.53',
       '0.98', '0.41', '3.54', '4.16', '6.04', '4.18', '3.84', '0.06',
       '0.47', '5.38', '5.32', '5.65', '1.87', '0.13', '3.12', '0.36',
       '0.11', '4.35', '0.65', '0.07', '0.08', '0.49', '0.3', '2.66',
       '2.69', '0.48', '0.38', '5.33', '1.91', '3.96', '3.1', '1.1',
       '1.2', '0.14', '2.54', '2.14', '0.81', '2.12', '0.44', '3.15',
       '1.25', '0.04', '0.0', '2.47', '2.23', '1.69', '0.01', '3.0',
       '0.02', '4.39', '1.98', '0.1', '3.81', '0.05', '2.49', '1.58',
       '3.14', '2.73', '0.66', '0.22', '3.63', '1.45', '1.31', '2.43',
       '0.7', '0.35', '1.4', '0.6', '2.26', '1.42', '1.28', '1.39',
       '0.87', '0.17', '0.94', '0.19', '0.21', '1.6', '0.16', '1.03',
       '0.25', '2.06', '1.49', '1.29', '0.09', '2.87', '0.03', '0.78',
       '0.83', '2.33', '2.02', '1.36', '1.81', '1.97', '0.91', '0.99',
       '0.95', '2.0'

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'],
      dtype=object)

array([76., nan, 82., 80., 89., 58., 87., 91., 61., 97., 95., 77., 88.,
       83., 94., 93., 85., 86., 98., 96., 90., 84., 73., 74., 78., 92.,
       71., 72., 68., 62., 49., 67., 81., 66., 56., 79., 70., 59., 64.,
       75., 60., 63., 69., 50., 25., 42., 44., 55., 48., 57., 29., 47.,
       65., 54., 20., 53., 37., 38., 33., 52., 30., 32., 43., 45., 51.,
       40., 46., 39., 34., 35., 41., 36., 28., 31., 27., 26., 19., 23.,
       24., 21., 17., 22., 13.])

array([2006., 1985., 2008., 2009., 1996., 1989., 1984., 2005., 1999.,
       2007., 2010., 2013., 2004., 1990., 1988., 2002., 2001., 2011.,
       1998., 2015., 2012., 2014., 1992., 1997., 1993., 1994., 1982.,
       2016., 2003., 1986., 2000.,   nan, 1995., 1991., 1981., 1987.,
       1980., 1983.])

array(['E', nan, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

Требуется преобразовать 'Year of Release', 'EU sales', 'JP sales', 'User Score', там некорректные типы данных

In [9]:

# Преобразование данных в числовой формат
df['year_of_release'] = pd.to_numeric(df['year_of_release'], errors='coerce').fillna(-1).astype('int64')
df['eu_sales'] = pd.to_numeric(df['eu_sales'], errors='coerce').fillna(-1).astype('float64')
df['jp_sales'] = pd.to_numeric(df['jp_sales'], errors='coerce').fillna(-1).astype('float64')
df['user_score'] = pd.to_numeric(df['user_score'], errors='coerce').fillna(-1).astype('float64')
df['critic_score'] = pd.to_numeric(df['critic_score'], errors='coerce').fillna(-1).astype('int64')
display(df)
df.info()

# Проверка структуры данных

name platform  year_of_release         genre  \
0                         Wii Sports      Wii             2006        Sports   
1                  Super Mario Bros.      NES             1985      Platform   
2                     Mario Kart Wii      Wii             2008        Racing   
3                  Wii Sports Resort      Wii             2009        Sports   
4           Pokemon Red/Pokemon Blue       GB             1996  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3             2016        Action   
16952               LMA Manager 2007     X360             2006        Sports   
16953        Haitaka no Psychedelica      PSV             2016     Adventure   
16954               Spirits & Spells      GBA             2003      Platform   
16955            Winning Post 8 2016      PSV             2016    Simulation   

       na_sales  eu_sales  jp_sales  other_sales  critic_score  user_score  \
0         41.36     28.96      3.77         8.45            76         8.0   
1         29.08      3.58      6.81         0.77            -1        -1.0   
2         15.68     12.76      3.79         3.29            82         8.3   
3         15.61     10.93      3.28         2.95            80         8.0   
4         11.27      8.89     10.22         1.00            -1        -1.0   
...         ...       ...       ...          ...           ...         ...   
16951      0.00      0.00      0.01         0.00            -1        -1.0   
16952      0.00      0.01      0.00         0.00            -1        -1.0   
16953      0.00      0.00      0.01         0.00            -1        -1.0   
16954      0.01      0.00      0.00         0.00            -1        -1.0   
16955      0.00      0.00      0.01         0.00            -1        -1.0   

      rating  
0          E  
1        NaN  
2          E  
3          E  
4        NaN  
...      ...  
16951    NaN  
16952    NaN  
16953    NaN  
16954    NaN  
16955    NaN  

[16956 rows x 11 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16956 non-null  int64  
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16956 non-null  float64
 6   jp_sales         16956 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     16956 non-null  int64  
 9   user_score       16956 non-null  float64
 10  rating           10085 non-null  object 
dtypes: float64(5), int64(2), object(4)
memory usage: 1.4+ MB


### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


In [10]:
# Подсчитываем количество пропусков в каждом столбце
missing_counts = df.isna().sum()

# Рассчитываем долю пропусков
missing_part = missing_counts / len(df)

# Создаем датафрейм с результатами
missing_data = pd.DataFrame({'Количество пропусков': missing_counts, 'Доля пропусков': missing_part})

# Выводим результаты
print(missing_data)

                 Количество пропусков  Доля пропусков
name                                2        0.000118
platform                            0        0.000000
year_of_release                     0        0.000000
genre                               2        0.000118
na_sales                            0        0.000000
eu_sales                            0        0.000000
jp_sales                            0        0.000000
other_sales                         0        0.000000
critic_score                        0        0.000000
user_score                          0        0.000000
rating                           6871        0.405225


Данные из пункта 1!!! Сейчас после преобразования данные изменились.

Самое большое количество пропусков хранится в столбцах 'Critic Score' - 51.4%, 'User Score' - 40.1%, 'Rating' - 40.5%. 
В столбце year_of_release число пропусков менее 2%, этими данными по годам можно пренебречь и удалить их, они не повлияют на результат. В столбцах critic_score, user_score и rating пропущенные значения составляют 40-50%. Вероятно, это связано с тем, что данные игры не были популярны, практически не пользовались спросом. Пропуски в столбцах 'eu_sales', 'jp_sales', critic_score, user_score заполним средними значениями. 'Rating' заполнил значением no_rating

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

In [11]:
# заполним пропуски средним значением в 'eu_sales' и 'jp_sales' с помощью функции:
def fill_sales(series):
    return series.fillna(series.mean())

# Применяем функцию к нужным столбцам
df['eu_sales'] = fill_sales(df['eu_sales'])
df['jp_sales'] = fill_sales(df['jp_sales'])

# Заполняем пропуски значениями -1 для critic_score и user_score, для последующей категоризации
# Заполнение поля rating значением no_rating
df['rating'] = df['rating'].fillna('no_rating')  
df['critic_score'] = df['critic_score'].fillna(-1)
df['user_score'] = df['user_score'].fillna(-1)
df

name platform  year_of_release         genre  \
0                         Wii Sports      Wii             2006        Sports   
1                  Super Mario Bros.      NES             1985      Platform   
2                     Mario Kart Wii      Wii             2008        Racing   
3                  Wii Sports Resort      Wii             2009        Sports   
4           Pokemon Red/Pokemon Blue       GB             1996  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3             2016        Action   
16952               LMA Manager 2007     X360             2006        Sports   
16953        Haitaka no Psychedelica      PSV             2016     Adventure   
16954               Spirits & Spells      GBA             2003      Platform   
16955            Winning Post 8 2016      PSV             2016    Simulation   

       na_sales  eu_sales  jp_sales  other_sales  critic_score  user_score  \
0         41.36     28.96      3.77         8.45            76         8.0   
1         29.08      3.58      6.81         0.77            -1        -1.0   
2         15.68     12.76      3.79         3.29            82         8.3   
3         15.61     10.93      3.28         2.95            80         8.0   
4         11.27      8.89     10.22         1.00            -1        -1.0   
...         ...       ...       ...          ...           ...         ...   
16951      0.00      0.00      0.01         0.00            -1        -1.0   
16952      0.00      0.01      0.00         0.00            -1        -1.0   
16953      0.00      0.00      0.01         0.00            -1        -1.0   
16954      0.01      0.00      0.00         0.00            -1        -1.0   
16955      0.00      0.00      0.01         0.00            -1        -1.0   

          rating  
0              E  
1      no_rating  
2              E  
3              E  
4      no_rating  
...          ...  
16951  no_rating  
16952  no_rating  
16953  no_rating  
16954  no_rating  
16955  no_rating  

[16956 rows x 11 columns]

In [12]:
# Удаляем строки с пропусками в year_of_release. Ихвсего 275 штук
df = df[~df['year_of_release'].isna()]

# Выводим обновленный датафрейм
df


name platform  year_of_release         genre  \
0                         Wii Sports      Wii             2006        Sports   
1                  Super Mario Bros.      NES             1985      Platform   
2                     Mario Kart Wii      Wii             2008        Racing   
3                  Wii Sports Resort      Wii             2009        Sports   
4           Pokemon Red/Pokemon Blue       GB             1996  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3             2016        Action   
16952               LMA Manager 2007     X360             2006        Sports   
16953        Haitaka no Psychedelica      PSV             2016     Adventure   
16954               Spirits & Spells      GBA             2003      Platform   
16955            Winning Post 8 2016      PSV             2016    Simulation   

       na_sales  eu_sales  jp_sales  other_sales  critic_score  user_score  \
0         41.36     28.96      3.77         8.45            76         8.0   
1         29.08      3.58      6.81         0.77            -1        -1.0   
2         15.68     12.76      3.79         3.29            82         8.3   
3         15.61     10.93      3.28         2.95            80         8.0   
4         11.27      8.89     10.22         1.00            -1        -1.0   
...         ...       ...       ...          ...           ...         ...   
16951      0.00      0.00      0.01         0.00            -1        -1.0   
16952      0.00      0.01      0.00         0.00            -1        -1.0   
16953      0.00      0.00      0.01         0.00            -1        -1.0   
16954      0.01      0.00      0.00         0.00            -1        -1.0   
16955      0.00      0.00      0.01         0.00            -1        -1.0   

          rating  
0              E  
1      no_rating  
2              E  
3              E  
4      no_rating  
...          ...  
16951  no_rating  
16952  no_rating  
16953  no_rating  
16954  no_rating  
16955  no_rating  

[16956 rows x 11 columns]

### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

In [13]:
# Уникальные значения в категориальных данных
unique_genres = df['genre'].unique()
unique_platforms = df['platform'].unique()
unique_ratings = df['rating'].unique()
unique_years = df['year_of_release'].unique()

display(unique_genres)
display(unique_platforms)
display(unique_ratings)
display(unique_years)


array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', nan, 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION',
       'SHOOTER', 'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE',
       'SIMULATION', 'PUZZLE', 'STRATEGY'], dtype=object)

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

array(['E', 'no_rating', 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'],
      dtype=object)

array([2006, 1985, 2008, 2009, 1996, 1989, 1984, 2005, 1999, 2007, 2010,
       2013, 2004, 1990, 1988, 2002, 2001, 2011, 1998, 2015, 2012, 2014,
       1992, 1997, 1993, 1994, 1982, 2016, 2003, 1986, 2000,   -1, 1995,
       1991, 1981, 1987, 1980, 1983])

In [14]:
# Меняем регистр 
df['genre'] = df['genre'].str.lower()
df['platform'] = df['platform'].str.lower()
df['rating'] = df['rating'].str.upper()



In [15]:
unique_combinations_count = df[['genre', 'platform', 'rating', 'year_of_release']].nunique()
# Выводим результат
display(f'Количество уникальных комбинаций жанр-платформа: {unique_combinations_count}')


'Количество уникальных комбинаций жанр-платформа: genre              12\nplatform           31\nrating              9\nyear_of_release    38\ndtype: int64'

- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [16]:
# Сортируем датафрейм по всем столбцам
df_sorted = df.sort_values(by=df.columns.tolist())

# Находим дубликаты
duplicates = df_sorted[df_sorted.duplicated(keep=False)]

# Выводим дубликаты
duplicates.head(10)

name platform  year_of_release  \
15191                             Beyblade Burst      3ds             2016   
15192                             Beyblade Burst      3ds             2016   
15301                          11eyes: CrossOver     x360             2009   
15302                          11eyes: CrossOver     x360             2009   
4860            18 Wheeler: American Pro Trucker      ps2             2001   
4861            18 Wheeler: American Pro Trucker      ps2             2001   
13098                                 4 Elements       pc             2009   
13099                                 4 Elements       pc             2009   
5235   999: Nine Hours, Nine Persons, Nine Doors       ds             2009   
5236   999: Nine Hours, Nine Persons, Nine Doors       ds             2009   

              genre  na_sales  eu_sales  jp_sales  other_sales  critic_score  \
15191  role-playing      0.00      0.00      0.03         0.00            -1   
15192  role-playing      0.00      0.00      0.03         0.00            -1   
15301     adventure      0.00      0.00      0.02         0.00            -1   
15302     adventure      0.00      0.00      0.02         0.00            -1   
4860         racing      0.20      0.15      0.00         0.05            61   
4861         racing      0.20      0.15      0.00         0.05            61   
13098        puzzle      0.00      0.04      0.00         0.01            -1   
13099        puzzle      0.00      0.04      0.00         0.01            -1   
5235      adventure      0.31      0.00      0.03         0.02            -1   
5236      adventure      0.31      0.00      0.03         0.02            -1   

       user_score     rating  
15191        -1.0  NO_RATING  
15192        -1.0  NO_RATING  
15301        -1.0  NO_RATING  
15302        -1.0  NO_RATING  
4860          5.7          E  
4861          5.7          E  
13098         7.4          E  
13099         7.4          E  
5235         -1.0  NO_RATING  
5236         -1.0  NO_RATING

In [17]:
# Определяем количество строк до удаления дубликатов
initial_row_count = len(df)
print(f'Количество строк до удаления дубликатов: {initial_row_count}')

# Находим дублирующиеся строки
duplicates = df[df.duplicated()]
print("Дублирующиеся строки:")
print(duplicates)

# Удаляем дублирующиеся строки
games_cleaned = df.drop_duplicates()

# Определяем количество строк после удаления дубликатов
final_row_count = len(games_cleaned)
print(f'Количество строк после удаления дубликатов: {final_row_count}')

Количество строк до удаления дубликатов: 16956
Дублирующиеся строки:
                                                name platform  \
268                            Batman: Arkham Asylum      ps3   
368                 James Bond 007: Agent Under Fire      ps2   
717                            God of War: Ascension      ps3   
823                                Wipeout: The Game      wii   
848                  Rayman Raving Rabbids: TV Party      wii   
...                                              ...      ...   
16671        Fullmetal Alchemist: Prince of the Dawn      wii   
16753                                      Routes PE      ps2   
16799                            Transformers: Prime      wii   
16912  Metal Gear Solid V: The Definitive Experience     xone   
16940                          The Longest 5 Minutes      psv   

       year_of_release      genre  na_sales  eu_sales  jp_sales  other_sales  \
268               2009     action      2.24      1.31      0.07       

- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

После обработки и нормализации данных были выведены строки до удаления дубликатов, найдены 241 явных дубликатов, они удалены.


Перед этим во время были удалены строки с пропущенными данными в year_of_release.

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

In [18]:
total_rows = 16956
final_row_count = 16956-241-275

deleted_rows = total_rows - final_row_count

deleted_share = round((deleted_rows / total_rows), 4)

display(deleted_rows)
display(deleted_share)

516

0.0304

- После проведения предобработки данных напишите общий промежуточный вывод.

В процессе предобработки были удалены ячейки с пропущенным годом выпуска (275 штук), а также 241 явных дубликатов, 
суммарно это 516 строк (3.04% от всех значений)

Доля ячеек с пропущенным годом выпуска: 275 / 16956 = 1.62%   
Доля ячеек с явными дубликатами: 241 / 16956 = 1.42%

Были проверены явные и неявные дубликаты. 
Ячейки 'year_of_release' и 'critic_score' приведены к типу int64, ячейки 'eu_sales', 'jp_sales', 'user_score' к типу float64.
Заполнены средние значения в 'eu_sales', 'jp_sales', а также -1 в 'user_score', 'critic_score', no_rating в rating.

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [19]:
df_actual = df.copy()
# Ставим диапазон от 2000 до 2013гг включительно
df_actual = df[(df['year_of_release'] >= 2000) & (df['year_of_release'] <= 2013)]

# Смотрим данные нового датафрейма
print(f'Количество строк в df_actual: {len(df_actual)}')
print(df_actual.info())

Количество строк в df_actual: 12980
<class 'pandas.core.frame.DataFrame'>
Int64Index: 12980 entries, 0 to 16954
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12980 non-null  object 
 1   platform         12980 non-null  object 
 2   year_of_release  12980 non-null  int64  
 3   genre            12980 non-null  object 
 4   na_sales         12980 non-null  float64
 5   eu_sales         12980 non-null  float64
 6   jp_sales         12980 non-null  float64
 7   other_sales      12980 non-null  float64
 8   critic_score     12980 non-null  int64  
 9   user_score       12980 non-null  float64
 10  rating           12980 non-null  object 
dtypes: float64(5), int64(2), object(4)
memory usage: 1.2+ MB
None


После фильтрации осталось 12980 строк в указанном диапазоне

In [20]:
# Всё это можно показать кодом (Смотри ниже)
 
 # код ревьюера
 # Создаем DataFrame
df_my = pd.DataFrame({
    'A': [1, 2, 3],
    'B': [4, 5, 6]
})

# Давай его посмотрим
print('Вот так выглядит df_my изначально')
display(df_my)

# А теперь давай создадим "копию" DataFrame без использования .copy()
df_fake_copy = df_my


# А теперь вносим изменения в "копию"
df_fake_copy.loc[0, 'A'] = 100
print('Вот так выглядит df_fake_copy  после внесения изменений')
display(df_fake_copy)

# А теперь вопрос на засыпку, что произойдёт с df_my?

 
 
 # Смотрим
display(df_my)
# И оказывается что он принял все изменения df_fake_copy


# Но в чём тогда смысл df_copy = df_my, если изменения в одном, немедленно приведёт к изменениям в другом? 
# 
# Чтобы сделать настоящую копию:

df_real_copy = df_my.copy()

# И теперь если мы внесём изменения в df_real_copy, это не приведёт к изменениям в df_my, а если мы изменим df_my, 
# df_real_copyэто не приведёт к изменениям 

#  обрати на это внимание

# Почему так происходит .copy()? Python, объекты, такие как DataFrame, являются ссылочными типами данных. 
# Это означает, что переменные, которые хранят эти объекты, фактически хранят ссылки на место в памяти, 
# где находится объект. Давай посмотрим id df_my и id df_fake_copy
df_my_id = id(df_my)
df_fake_copy_id = id(df_fake_copy)

print(f"Смотрим сектора хранения информации df_my: {df_my_id} и df_fake_copy_id: {df_fake_copy_id}")

# И видим что они идентичные! То есть наши это фреймы по-прежнему ссылаются на один и тот же сектор
# Вот поэтому изменения в одном приводят к изменению в другом

Вот так выглядит df_my изначально


A  B
0  1  4
1  2  5
2  3  6

Вот так выглядит df_fake_copy  после внесения изменений


A  B
0  100  4
1    2  5
2    3  6

A  B
0  100  4
1    2  5
2    3  6

Смотрим сектора хранения информации df_my: 140629824053552 и df_fake_copy_id: 140629824053552


---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [21]:


# Создаем столбец с категориями на основе пользовательских оценок
def user_category_score(score):
    if score >= 8 and score <= 10:
        return 'высокая'
    elif score >= 3 and score < 8:
        return 'средняя'
    elif score >= 0 and score < 3:
        return 'низкая'
    else:
        return 'не определено'

df['user_category'] = df['user_score'].apply(user_category_score)

# Выводим результат
df[['user_score', 'user_category']].head()



user_score  user_category
0         8.0        высокая
1        -1.0  не определено
2         8.3        высокая
3         8.0        высокая
4        -1.0  не определено

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [22]:

# Создаем столбец с категориями на основе оценок критиков
def critic_category_score(score):
    if score >= 80 and score <= 100:
        return 'высокая'
    elif score >= 30 and score < 80:
        return 'средняя'
    elif score >= 0 and score < 30:
        return 'низкая'
    else:
        return 'не определено'

df['critic_category'] = df['critic_score'].apply(critic_category_score)

# Выводим результат
df[['critic_score', 'critic_category']].head()


critic_score critic_category
0            76         средняя
1            -1   не определено
2            82         высокая
3            80         высокая
4            -1   не определено

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [23]:
# Общее количество игр в разрезе оценки пользователя
total_by_user_score = df.groupby('user_category')['user_score'].count()
display('Общее количество игр по оценке пользователей:')
display(total_by_user_score)

'Общее количество игр по оценке пользователей:'

user_category
высокая          2610
не определено    9268
низкая            146
средняя          4932
Name: user_score, dtype: int64

In [24]:
# Общее количество игр в разрезе оценки критика
total_by_critic_score = df.groupby('critic_category')['critic_score'].count()
display('Общее количество игр по оценке критиков:')
display(total_by_critic_score)

'Общее количество игр по оценке критиков:'

critic_category
высокая          2033
не определено    8714
низкая             62
средняя          6147
Name: critic_score, dtype: int64

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [25]:
platform_counts = df_actual['platform'].value_counts()

top7_platforms = platform_counts.head(7)
top7_platforms

ps2     2154
ds      2146
wii     1294
psp     1199
x360    1138
ps3     1107
gba      826
Name: platform, dtype: int64

---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

В проекте первоначально загрузили данные и ознакомились с ними. Выяснили исходные данные, количество строк и столбцов.
Изучили долю пропусков в столбцах. Самое большое количество пропусков было в столбцах "оценка пользователя", "оценка критика" и "рейтинг" (40-50%).
Год привел к целочисленному значению, оценку пользователей - к вещественному. Долю продаж в Японии и Европе - также к вещественному.
Заменил пропуски в "оценка пользователя" и "оценка критика" на среднее значение для корректных расчетов, в поле "рейтинг" указал отсутствие рейтинга.
Пропуски в поле "год" удалил, так как их доля менее 2% и они не сыграют существенную роль.
Привел наименования столбцов и значения к нижнему регистру.
Удалил явные дубликаты. Суммарно доля удаленных щначений составила 516 (3.04% от общего количества). Это 241 явных дубликатов и 275 полей с пропущенным годом.
Были отфильрованы данные по годам выпуска с 2000 по 2013, также была проведена категоризация по оценкам пользователей и критиков, а также выведены топ-7 платформ по количеству игр.




Самое большое количество игр было на платформе PS2 (2154), чуть менее у DS (2146), далее Wii (1199).
Абсолютное большинство игр получили средние оценки критиков (от 30 до 80) и оценки пользователей (от 3 до 8).
Количество игр с 2000 по 2013 год составило 12980 штук (всего в таблице 16956).
Доля удаленных строк составила 3% - это явные дубликаты и пропуски заполненных лет.
В таблице представлены данные с 1980 по 2016 год (36 лет), при этом за проанализированные 13 лет (36% от количества взятых лет) было выпущено 76% игр (12980 / 16956).
Это связано с бурным развитием компьютерных технологий и игр в период с 2000 по 2013 гг.
